cg/server/api.py

```diff
@@ -30,54 +36,25 @@
     SampleLaneSequencingMetrics,
     User,
 )
-from flask import Blueprint, abort, current_app, g, jsonify, make_response, request
-from google.auth import jwt
-from pydantic.v1 import ValidationError
-from requests.exceptions import HTTPError
-from sqlalchemy.exc import IntegrityError
-from urllib3.exceptions import MaxRetryError, NewConnectionError
-from werkzeug.utils import secure_filename
 
 LOG = logging.getLogger(__name__)
 BLUEPRINT = Blueprint("api", __name__, url_prefix="/api/v1")
 
-cache = TTLCache(maxsize=1, ttl=3600)
-cache_certificates_key = "certs"
-cache[cache_certificates_key] = None
-
-
-def get_certificate_ttl(response_data) -> int:
-    """Extract time to live in seconds for certificate from response headers."""
-    expires_header = response_data.headers.get("Expires")
-    expires = datetime.strptime(expires_header, "%a, %d %b %Y %H:%M:%S %Z")
-    return int((expires - datetime.utcnow()).total_seconds())
-
-
-def fetch_and_cache_google_oauth2_certificates():
-    """Fetch and cache Google OAuth2 certificates."""
-    global cache
 
-    url = "https://www.googleapis.com/oauth2/v1/certs"
-    response = requests.get(url)
-    response.raise_for_status()
+def verify_google_token(token):
+    """Validate expiration, issuer, audience and signature of a Google OAuth2 token."""
+    client_id: str = current_app.config["GOOGLE_OAUTH_CLIENT_ID"]
 
-    certs = response.json()
-    ttl = get_certificate_ttl(response_data=response)
-
-    cache = TTLCache(maxsize=1, ttl=ttl)
-    cache[cache_certificates_key] = certs
-
-    return certs
-
-
-def get_google_oauth2_certificates():
-    """Get Google OAuth2 certificates from cache or fetch and cache them."""
-    certs = cache.get(cache_certificates_key)
+    user_data = id_token.verify_oauth2_token(
+        id_token=token, request=google_requests.Request(), audience=client_id
+    )
+    issuer: str = user_data["iss"]
+    valid_issuers = ["accounts.google.com", "https://accounts.google.com"]
 
-    if not certs:
-        certs = fetch_and_cache_google_oauth2_certificates()
+    if issuer not in valid_issuers:
+        raise ValueError("Invalid issuer")
```

Comment:
I think this check is already performed in verify_oath2_token: https://google-auth.readthedocs.io/en/stable/_modules/google/oauth2/id_token.html#verify_oauth2_token

Reply:
Great! Nice 😄